Mathlib/Data/Polynomial/Derivation.lean

```diff
@@ -0,0 +1,70 @@
+/-
+Copyright (c) 2023 Kevin Buzzard. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Buzzard, Richard Hill
+-/
+import Mathlib.Data.Polynomial.Derivative
+import Mathlib.RingTheory.Derivation.Basic
+import Mathlib.Data.Polynomial.AlgebraMap
+
+/-!
+# Derivations of univariate polynomials
+
+In this file we prove that an `R`-derivation of `Polynomial R` is determined by its value on `X`.
+We also provide a constructor `Polynomial.mkDerivation` that
+builds a derivation from its value on `X`, and a linear equivalence
+`Polynomial.mkDerivationEquiv` between `A` and `Derivation (Polynomial R) A`.
+-/
+
+namespace Polynomial
+
+open scoped BigOperators
+
+noncomputable section
+
+variable {R A : Type _} [CommSemiring R] [AddCommMonoid A] [Module R A] [Module (Polynomial R) A]
+
+section
+
+@[simp]
+theorem derivation_C (D : Derivation R (Polynomial R) A) (a : R) : D (C a) = 0 :=
+  D.map_algebraMap a
+
+-- simp normal form is `derivation_C_mul'`
+theorem derivation_C_mul (D : Derivation R (Polynomial R) A) (a : R) (f : Polynomial R) :
+    D (C a * f) = a • D f := by rw [C_mul', D.map_smul]
+
+@[simp]
+theorem derivation_C_mul' (D : Derivation R (Polynomial R) A) (a : R) (f : Polynomial R) :
+    C a • D f = a • D f := by
+  have : C a • D f = D (C a * f) := by simp
+  rw [this, derivation_C_mul]
+
+@[ext]
+theorem derivation_ext {D₁ D₂ : Derivation R (Polynomial R) A} (h : D₁ X = D₂ X) : D₁ = D₂ :=
+  Derivation.ext fun f => Derivation.eqOn_adjoin (Set.eqOn_singleton.2 h) (mem_adjoin_X f)
+
+variable [IsScalarTower R (Polynomial R) A]
+
+variable (R)
+
+/-- The derivation on `R[X]` that takes the value `a` on `X`. -/
+def mkDerivation (a : A) : Derivation R (Polynomial R) A where
+  toLinearMap := (LinearMap.restrictScalars R <| LinearMap.toSpanSingleton R[X] A a).comp derivative
+  map_one_eq_zero' := by simp
+  leibniz' := by
+    intros f g
+    simp [mul_comm, mul_smul, add_comm]
+
+@[simp]
+theorem mkDerivation_X (a : A) : mkDerivation R a X = a := by simp [mkDerivation]
+
+/-- `Polynomial.mkDerivation` as a linear equivalence. -/
+def mkDerivationEquiv : A ≃ₗ[R] Derivation R R[X] A :=
```

Comment:
```suggestion
@[simps]
def mkDerivationEquiv : A ≃ₗ[R] Derivation R R[X] A :=
```

Reply:
Will `@[simps!]` do? Asking for a linter. The reason we have a `LinearEquiv.symm` term here is that `map_add'` and `map_smul'` are much easier that way around (they're `rfl`). Looking at the output with `@[simps!?]` most of it looks OK, apart from
```
[simps.verbose] adding projection Polynomial.mkDerivationEquiv_symm_apply:
      ∀ (R : Type u_1) {A : Type u_2} [inst : CommSemiring R] [inst_1 : AddCommMonoid A] [inst_2 : Module R A]
        [inst_3 : Module R[X] A] [inst_4 : IsScalarTower R R[X] A] (a : Derivation R R[X] A),
        ↑(LinearEquiv.symm (Polynomial.mkDerivationEquiv R)) a =
          ↑(↑(LinearEquiv.symm
                    {
                      toLinearMap :=
                        {
                          toAddHom :=
                            { toFun := fun D ↦ ↑D Polynomial.X,
                              map_add' :=
                                (_ :
                                  ∀ (x x_1 : Derivation R R[X] A),
                                    (fun D ↦ ↑D Polynomial.X) (x + x_1) = (fun D ↦ ↑D Polynomial.X) (x + x_1)) },
                          map_smul' :=
                            (_ :
                              ∀ (x : R) (x_1 : Derivation R R[X] A),
                                AddHom.toFun
                                    { toFun := fun D ↦ ↑D Polynomial.X,
                                      map_add' :=
                                        (_ :
                                          ∀ (x x_2 : Derivation R R[X] A),
                                            (fun D ↦ ↑D Polynomial.X) (x + x_2) = (fun D ↦ ↑D Polynomial.X) (x + x_2)) }
                                    (x • x_1) =
                                  AddHom.toFun
                                    { toFun := fun D ↦ ↑D Polynomial.X,
                                      map_add' :=
                                        (_ :
                                          ∀ (x x_2 : Derivation R R[X] A),
                                            (fun D ↦ ↑D Polynomial.X) (x + x_2) = (fun D ↦ ↑D Polynomial.X) (x + x_2)) }
                                    (x • x_1)) },
                      invFun := Polynomial.mkDerivation R,
                      left_inv :=
                        (_ :
                          ∀ (x : Derivation R R[X] A),
                            Polynomial.mkDerivation R
                                (AddHom.toFun
                                  {
                                      toAddHom :=
                                        { toFun := fun D ↦ ↑D Polynomial.X,
                                          map_add' :=
                                            (_ :
                                              ∀ (x x_1 : Derivation R R[X] A),
                                                (fun D ↦ ↑D Polynomial.X) (x + x_1) =
                                                  (fun D ↦ ↑D Polynomial.X) (x + x_1)) },
                                      map_smul' :=
                                        (_ :
                                          ∀ (x : R) (x_1 : Derivation R R[X] A),
                                            AddHom.toFun
                                                { toFun := fun D ↦ ↑D Polynomial.X,
                                                  map_add' :=
                                                    (_ :
                                                      ∀ (x x_2 : Derivation R R[X] A),
                                                        (fun D ↦ ↑D Polynomial.X) (x + x_2) =
                                                          (fun D ↦ ↑D Polynomial.X) (x + x_2)) }
                                                (x • x_1) =
                                              AddHom.toFun
                                                { toFun := fun D ↦ ↑D Polynomial.X,
                                                  map_add' :=
                                                    (_ :
                                                      ∀ (x x_2 : Derivation R R[X] A),
                                                        (fun D ↦ ↑D Polynomial.X) (x + x_2) =
                                                          (fun D ↦ ↑D Polynomial.X) (x + x_2)) }
                                                (x • x_1)) }.toAddHom
                                  x) =
                              x),
                      right_inv := (_ : ∀ (x : A), ↑(Polynomial.mkDerivation R x) Polynomial.X = x) })).symm
            a 
```
I don't really understand `simps` properly. Should I tell simps somehow to not generate this lemma and then write a more appropriate one by hand?

Reply:
There used to be a `simp_rhs := true` setting in Lean 3. I don't know if that's what `!` means (in which case, there's a simp lemma missing), or if you have to spell it differently.

I think it's fine to write the lemma manually if `simps` is generating a mess.

Reply:
OK having read the docstring I understand `simps` much better. I think I'll just write these manually, given that as far as I can see there are only two.